esp-hal-common/src/uart.rs

```diff
@@ -1442,67 +1484,66 @@ mod asynch {
     where
         T: Instance,
     {
-        /// Read async to buffer slice `buf`. Wait for Rx Fifo Full interrupt
-        /// (set by `set_rx_fifo_full_threshold`) and/or Rx AT_CMD character
-        /// interrupt if `set_at_cmd` was called.
+        /// Read async to buffer slice `buf`.
+        /// Waits until one of the following interrupts occurs:
+        /// - `RXFIFO_FULL`
+        /// - `RXFIFO_OVF`
+        /// - `AT_CMD_CHAR_DET` (only if `set_at_cmd` was called)
+        /// - `RXFIFO_TOUT` (only if `set_rx_timeout was called)
+        ///
+        /// The interrupts in question are enabled during the body of this
+        /// function. The method immediately returns when the interrupt
+        /// has already occured before calling this method (e.g. status
+        /// bit set, but interrupt not enabled)
         ///
         /// # Params
         /// - `buf` buffer slice to write the bytes into
         ///
         /// # Errors
         /// - `Err(RxFifoOvf)` when MCU Rx Fifo Overflow interrupt is triggered.
         ///   To avoid this error, call this function more often.
-        /// - `Err(Error::ReadNoConfig)` if neither `set_rx_fifo_full_threshold`
-        ///   or `set_at_cmd` was called
         ///
         /// # Ok
-        /// When succesfull, returns the number of bytes written to
-        /// buf
+        /// When succesfull, returns the number of bytes written to buf
         async fn read_async(&mut self, buf: &mut [u8]) -> Result<usize, Error> {
             if buf.len() == 0 {
                 return Ok(0);
             }
 
-            let mut read_bytes = 0;
+            let mut interrupts_to_await = UartInterrupt::RXFIFO_FULL | UartInterrupt::RXFIFO_OVF;
 
             if self.at_cmd_config.is_some() {
-                if let Either3::Third(_) = select3(
```

Comment:
We can clean this `select` up by using [select_slice](https://docs.rs/embassy-futures/latest/embassy_futures/select/fn.select_slice.html). We can conditionally add to the slice of futures to await based on certain preconditions, for example, we would only add `UartFuture::<T>::new(Event::RxCmdCharDetected)` to the slice if `self.at_cmd_config.is_some()` is true.

Reply:
I thought about this approach too, but then I need something like heapless::Vector in order to generate the slice right?

Reply:
We could either do it manually by managing our slice of futures or pull in heapless. Personally, I don't mind adding heapless as a dep.